File: sdk/node/protocol.mjs
/** MirrorGate v1 public protocol. See protocol/ and shared conformance vectors. */
export const LIMITS = Object.freeze({frameBytes: 65535, manifestBytes: 262144, depth: 96, nodes: 8192, semanticDepth: 32});
const ID = /^[A-Za-z][A-Za-z0-9_.-]{0,127}$/;
const INTEGER = /^(?:0|-?[1-9][0-9]*)$/;
const DIGEST = /^[0-9a-f]{64}$/;
export class ProtocolError extends Error {
  constructor(code, message) { super(message); this.name = 'ProtocolError'; this.code = code; }
}
const bad = (message, code = 'SCHEMA') => { throw new ProtocolError(code, message); };
export function object(value) { return value !== null && typeof value === 'object' && !Array.isArray(value); }
export function exact(value, keys, code = 'SCHEMA') {
  if (!object(value) || Object.keys(value).length !== keys.length || keys.some(key => !Object.hasOwn(value, key))) bad('Unexpected object fields', code);
}
function scalarString(value) { return typeof value === 'string' && !/[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?<![\uD800-\uDBFF])[\uDC00-\uDFFF]/u.test(value); }
function name(value) { if (!scalarString(value) || Buffer.byteLength(value) < 1 || Buffer.byteLength(value) > 128) bad('Invalid public field name'); }
function id(value) { if (typeof value !== 'string' || !ID.test(value)) bad('Invalid stable identifier'); }
function requestId(value) { if (!Number.isSafeInteger(value) || value <= 0) bad('Invalid request ID'); }
function ownObject(entries) { return Object.fromEntries(entries); }

/** JSON.parse accepts duplicate keys; this bounded parser deliberately does not. */
export function parseJson(text, maxBytes = LIMITS.frameBytes) {
  if (typeof text !== 'string' || Buffer.byteLength(text) > maxBytes) bad('JSON byte limit exceeded', 'LIMIT');
  let pos = 0; let nodes = 0;
  const ws = () => { while (pos < text.length && /[ \t\r\n]/.test(text[pos])) pos++; };
  const string = () => {
    const start = pos++;
    while (pos < text.length) {
      const c = text[pos++];
      if (c === '"') {
        let result;
        try { result = JSON.parse(text.slice(start, pos)); } catch { bad('Malformed JSON string', 'FRAME'); }
        if (!scalarString(result)) bad('Non-scalar Unicode string', 'FRAME');
        return result;
      }
      if (c === '\\') pos++;
    }
    bad('Unterminated JSON string', 'FRAME');
  };
  const value = depth => {
    if (depth > LIMITS.depth || ++nodes > LIMITS.nodes) bad('JSON structural limit exceeded', 'LIMIT');
    ws(); const c = text[pos];
    if (c === '"') return string();
    if (c === '{') {
      pos++; ws(); const entries = []; const seen = new Set();
      if (text[pos] === '}') { pos++; return {}; }
      while (pos < text.length) {
        ws(); if (text[pos] !== '"') bad('Expected object key', 'FRAME');
        const key = string(); if (seen.has(key)) bad('Duplicate JSON object key', 'FRAME'); seen.add(key);
        ws(); if (text[pos++] !== ':') bad('Expected colon', 'FRAME');
        entries.push([key, value(depth + 1)]); ws();
        const sep = text[pos++]; if (sep === '}') return ownObject(entries);
        if (sep !== ',') bad('Expected object separator', 'FRAME');
      }
      bad('Unterminated JSON object', 'FRAME');
    }
    if (c === '[') {
      pos++; ws(); const result = [];
      if (text[pos] === ']') { pos++; return result; }
      while (pos < text.length) {
        result.push(value(depth + 1)); ws(); const sep = text[pos++];
        if (sep === ']') return result;
        if (sep !== ',') bad('Expected array separator', 'FRAME');
      }
      bad('Unterminated JSON array', 'FRAME');
    }
    for (const [literal, result] of [['true', true], ['false', false], ['null', null]]) {
      if (text.startsWith(literal, pos)) { pos += literal.length; return result; }
    }
    const matched = /^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?/.exec(text.slice(pos));
    if (matched) {
      const token = matched[0]; pos += token.length;
      const parts = /^(-?)([0-9]+)(?:\.([0-9]+))?(?:[eE]([+-]?[0-9]+))?$/.exec(token);
      let digits = (parts[2] + (parts[3] ?? '')).replace(/^0+/, '');
      const scale = Number(parts[4] ?? '0') - (parts[3]?.length ?? 0);
      if (!digits) return 0;
      if (!Number.isSafeInteger(scale)) bad('Unsafe JSON number', 'FRAME');
      if (scale < 0) {
        const remove = -scale;
        if (remove >= digits.length || !digits.endsWith('0'.repeat(Math.min(remove, digits.length)))) bad('Fractional JSON number', 'FRAME');
        digits = digits.slice(0, digits.length - remove);
      } else {
        if (digits.length + scale > 16) bad('Unsafe JSON number', 'FRAME');
        digits += '0'.repeat(scale);
      }
      const n = Number(parts[1] + digits);
      if (!Number.isSafeInteger(n)) bad('Unsafe JSON number', 'FRAME'); return n;
    }
    bad('Malformed JSON value', 'FRAME');
  };
  const result = value(0); ws(); if (pos !== text.length) bad('Trailing JSON content', 'FRAME'); return result;
}

export class FrameDecoder {
  constructor(onMessage) { this.buffer = Buffer.alloc(0); this.onMessage = onMessage; }
  push(chunk) {
    const data = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
    let start = 0;
    while (start < data.length) {
      const end = data.indexOf(10, start); const part = data.subarray(start, end < 0 ? data.length : end);
      if (this.buffer.length + part.length > LIMITS.frameBytes) bad('Frame byte limit exceeded', 'LIMIT');
      this.buffer = Buffer.concat([this.buffer, part]);
      if (end < 0) return;
      if (this.buffer.includes(13)) bad('CR is forbidden in JSONL frames', 'FRAME');
      let line; try { line = new TextDecoder('utf-8', {fatal: true, ignoreBOM: true}).decode(this.buffer); } catch { bad('Invalid UTF-8', 'FRAME'); }
      this.buffer = Buffer.alloc(0); const message = parseJson(line);
      if (!object(message)) bad('A frame must contain an object', 'FRAME');
      this.onMessage(message); start = end + 1;
    }
  }
  end() { if (this.buffer.length) bad('Unterminated JSONL frame', 'FRAME'); }
}
export function frame(value) {
  const output = JSON.stringify(value);
  if (Buffer.byteLength(output) > LIMITS.frameBytes) bad('Frame byte limit exceeded', 'LIMIT');
  // Apply the same structural bounds to output as input.
  if (!object(parseJson(output))) bad('A frame must contain an object', 'FRAME');
  return `${output}\n`;
}

export function validateType(type, depth = 0) {
  if (depth > LIMITS.semanticDepth) bad('Type depth limit exceeded', 'LIMIT');
  if (!object(type)) bad('Invalid public type');
  const child = t => validateType(t, depth + 1);
  switch (type.kind) {
    case 'int': case 'bool': case 'str': case 'null': exact(type, ['kind']); break;
    case 'seq': case 'set': exact(type, ['kind', 'element']); child(type.element); break;
    case 'tuple':
      exact(type, ['kind', 'elements']); if (!Array.isArray(type.elements)) bad('Invalid tuple elements'); type.elements.forEach(child); break;
    case 'record': {
      exact(type, ['kind', 'fields']); if (!Array.isArray(type.fields)) bad('Invalid record fields'); const seen = new Set();
      for (const field of type.fields) { exact(field, ['wireName', 'type']); name(field.wireName); if (seen.has(field.wireName)) bad('Duplicate record field'); seen.add(field.wireName); child(field.type); } break;
    }
    case 'map': exact(type, ['kind', 'key', 'value']); exact(type.key, ['kind']); if (type.key.kind !== 'str') bad('Only string map keys are portable'); child(type.value); break;
    case 'variant': {
      exact(type, ['kind', 'cases']); if (!Array.isArray(type.cases) || type.cases.length === 0) bad('Invalid variant cases'); const seen = new Set();
      for (const c of type.cases) { exact(c, ['tag', 'payload']); name(c.tag); if (seen.has(c.tag)) bad('Duplicate variant tag'); seen.add(c.tag); child(c.payload); } break;
    }
    default: bad('Unsupported public type');
  }
  return type;
}
export function validateManifest(manifest) {
  // Enforce object/aggregate bounds even for evaluator-owned in-memory values.
  parseJson(JSON.stringify(manifest), LIMITS.manifestBytes);
  exact(manifest, ['schema', 'interfaceDigest', 'initializers', 'actions', 'observations']);
  if (manifest.schema !== 'mirrorgate.port/v1' || !DIGEST.test(manifest.interfaceDigest)) bad('Invalid manifest identity');
  if (!Array.isArray(manifest.initializers) || manifest.initializers.length === 0 || !Array.isArray(manifest.actions) || !Array.isArray(manifest.observations) || manifest.observations.length === 0) bad('Invalid public port collections');
  const operations = new Set();
  for (const action of [...manifest.initializers, ...manifest.actions]) {
    exact(action, ['id', 'inputs']); id(action.id); if (operations.has(action.id)) bad('Duplicate operation ID'); operations.add(action.id);
    if (!Array.isArray(action.inputs)) bad('Invalid input collection'); const inputs = new Set();
    for (const input of action.inputs) { exact(input, ['id', 'type']); id(input.id); if (inputs.has(input.id)) bad('Duplicate input ID'); inputs.add(input.id); validateType(input.type); }
  }
  const outputs = new Set();
  for (const observation of manifest.observations) { exact(observation, ['id', 'type']); id(observation.id); if (outputs.has(observation.id)) bad('Duplicate observation ID'); outputs.add(observation.id); validateType(observation.type); }
  return manifest;
}
export function validateRequest(req) {
  if (!object(req) || req.v !== 1) bad('Unsupported protocol version'); requestId(req.id);
  switch (req.op) {
    case 'hello': exact(req, ['v', 'id', 'op', 'interfaceDigest', 'runtime']); if (!DIGEST.test(req.interfaceDigest)) bad('Invalid interface digest'); id(req.runtime); break;
    case 'create': case 'observe': case 'dispose': exact(req, ['v', 'id', 'op']); break;
    case 'invoke': exact(req, ['v', 'id', 'op', 'action', 'inputs']); id(req.action); if (!object(req.inputs)) bad('Invalid inputs envelope'); break;
    case 'cancel': exact(req, ['v', 'id', 'op', 'requestId']); requestId(req.requestId); if (req.requestId >= req.id) bad('Cancellation must target an earlier request'); break;
    default: bad('Unknown request operation');
  }
  return req;
}
export function validateResponse(reply) {
  if (!object(reply) || reply.v !== 1) bad('Unsupported response version'); requestId(reply.id);
  if (reply.ok === true) exact(reply, ['v', 'id', 'ok', 'result']);
  else if (reply.ok === false) { exact(reply, ['v', 'id', 'ok', 'error']); exact(reply.error, ['code', 'message']); if (typeof reply.error.code !== 'string' || !/^[A-Z][A-Z0-9_]{0,63}$/.test(reply.error.code)) bad('Invalid error code'); if (!scalarString(reply.error.message) || Buffer.byteLength(reply.error.message) > 1024) bad('Invalid error message'); }
  else bad('Invalid response status');
  return reply;
}
function canonical(type, value) {
  switch (type.kind) {
    case 'seq': return JSON.stringify(value.map(v => canonical(type.element, v)));
    case 'set': return JSON.stringify(value['#set'].map(v => canonical(type.element, v)).sort());
    case 'tuple': return JSON.stringify(type.elements.map((t, i) => canonical(t, value['#tup'][i])));
    case 'record': return JSON.stringify([...type.fields].sort((a, b) => a.wireName < b.wireName ? -1 : a.wireName > b.wireName ? 1 : 0).map(f => [f.wireName, canonical(f.type, value[f.wireName])]));
    case 'map': return JSON.stringify(value['#map'].map(([k, v]) => [k, canonical(type.value, v)]).sort((a, b) => a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0));
    case 'variant': return JSON.stringify([value.tag, canonical(type.cases.find(c => c.tag === value.tag).payload, value.value)]);
    default: return JSON.stringify(value);
  }
}
function duplicates(type, values) { const seen = new Set(); for (const item of values) { const key = canonical(type, item); if (seen.has(key)) bad('Duplicate collection element', 'VALUE'); seen.add(key); } }
function valueDepth(depth) { if (depth > LIMITS.semanticDepth) bad('Value depth limit exceeded', 'LIMIT'); }
export function decodeValue(type, value, depth = 0) {
  valueDepth(depth); const child = (t, v) => decodeValue(t, v, depth + 1);
  switch (type.kind) {
    case 'int': exact(value, ['#bigint'], 'VALUE'); if (typeof value['#bigint'] !== 'string' || !INTEGER.test(value['#bigint'])) bad('Invalid integer encoding', 'VALUE'); return BigInt(value['#bigint']);
    case 'bool': if (typeof value !== 'boolean') bad('Expected boolean', 'VALUE'); return value;
    case 'str': if (!scalarString(value)) bad('Expected Unicode string', 'VALUE'); return value;
    case 'null': if (value !== null) bad('Expected null', 'VALUE'); return null;
    case 'seq': if (!Array.isArray(value)) bad('Expected sequence', 'VALUE'); return Array.from(value, v => child(type.element, v));
    case 'set': exact(value, ['#set'], 'VALUE'); if (!Array.isArray(value['#set'])) bad('Expected set elements', 'VALUE'); { const decoded = value['#set'].map(v => child(type.element, v)); duplicates(type.element, value['#set']); return new Set(decoded); }
    case 'tuple': exact(value, ['#tup'], 'VALUE'); if (!Array.isArray(value['#tup']) || value['#tup'].length !== type.elements.length) bad('Tuple arity mismatch', 'VALUE'); return type.elements.map((t, i) => child(t, value['#tup'][i]));
    case 'record': exact(value, type.fields.map(f => f.wireName), 'VALUE'); return ownObject(type.fields.map(f => [f.wireName, child(f.type, value[f.wireName])]));
    case 'map': {
      exact(value, ['#map'], 'VALUE'); if (!Array.isArray(value['#map'])) bad('Expected map entries', 'VALUE'); const result = new Map();
      for (const entry of value['#map']) { if (!Array.isArray(entry) || entry.length !== 2 || !scalarString(entry[0]) || result.has(entry[0])) bad('Invalid or duplicate map key', 'VALUE'); result.set(entry[0], child(type.value, entry[1])); } return result;
    }
    case 'variant': { exact(value, ['tag', 'value'], 'VALUE'); const c = type.cases.find(c => c.tag === value.tag); if (!c) bad('Unknown variant tag', 'VALUE'); return {tag: c.tag, value: child(c.payload, value.value)}; }
    default: bad('Unsupported value type', 'VALUE');
  }
}
export function encodeValue(type, value, depth = 0) {
  valueDepth(depth); const child = (t, v) => encodeValue(t, v, depth + 1);
  switch (type.kind) {
    case 'int': if (typeof value !== 'bigint') bad('Expected native bigint', 'VALUE'); return {'#bigint': value.toString()};
    case 'bool': case 'str': case 'null': decodeValue(type, value, depth); return value;
    case 'seq': if (!Array.isArray(value)) bad('Expected native sequence', 'VALUE'); return Array.from(value, v => child(type.element, v));
    case 'tuple': if (!Array.isArray(value) || value.length !== type.elements.length) bad('Expected native tuple', 'VALUE'); return {'#tup': type.elements.map((t, i) => child(t, value[i]))};
    case 'set': if (!(value instanceof Set)) bad('Expected native Set', 'VALUE'); { const encoded = [...value].map(v => child(type.element, v)); duplicates(type.element, encoded); return {'#set': encoded}; }
    case 'record': exact(value, type.fields.map(f => f.wireName), 'VALUE'); return ownObject(type.fields.map(f => [f.wireName, child(f.type, value[f.wireName])]));
    case 'map': if (!(value instanceof Map)) bad('Expected native Map', 'VALUE'); return {'#map': [...value].map(([k, v]) => { if (!scalarString(k)) bad('Expected string map key', 'VALUE'); return [k, child(type.value, v)]; })};
    case 'variant': { exact(value, ['tag', 'value'], 'VALUE'); const c = type.cases.find(c => c.tag === value.tag); if (!c) bad('Unknown variant tag', 'VALUE'); return {tag: c.tag, value: child(c.payload, value.value)}; }
    default: bad('Unsupported native type', 'VALUE');
  }
}
export function decodeFields(fields, value) { exact(value, fields.map(f => f.id), 'VALUE'); return ownObject(fields.map(f => [f.id, decodeValue(f.type, value[f.id])])); }
export function encodeFields(fields, value) { exact(value, fields.map(f => f.id), 'VALUE'); return ownObject(fields.map(f => [f.id, encodeValue(f.type, value[f.id])])); }

File: supervisor/mirrorgate/protocol.py
"""Strict, language-neutral MirrorGate port/v1 admission and value validation.

This module does not load adapters, select sandbox permissions, or trust a
worker-supplied identity. See docs/protocol-v1.md for correlation and lifecycle.
"""

from __future__ import annotations

import json
import re
from decimal import Decimal, DecimalException
from pathlib import Path
from typing import Any

MAX_FRAME_BYTES = 65_535
MAX_MANIFEST_BYTES = 262_144
MAX_JSON_DEPTH = 96
MAX_TYPE_DEPTH = 32
MAX_VALUE_DEPTH = 32
MAX_NODES = 8_192
MAX_SAFE_INTEGER = 9_007_199_254_740_991
IDENTIFIER = re.compile(r"[A-Za-z][A-Za-z0-9_.-]{0,127}\Z", re.ASCII)
DIGEST = re.compile(r"[0-9a-f]{64}\Z", re.ASCII)
ERROR_CODE = re.compile(r"[A-Z][A-Z0-9_]{0,63}\Z", re.ASCII)
INTEGER = re.compile(r"(?:0|-?[1-9][0-9]*)\Z", re.ASCII)


class ProtocolError(ValueError):
    """Bounded public failure; no private evaluator content belongs here."""

    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code
        self.message = message


def _fail(code: str, message: str) -> None:
    raise ProtocolError(code, message)


def _string(value: Any, maximum: int | None = None, nonempty: bool = False) -> None:
    if not isinstance(value, str):
        _fail("SCHEMA", "expected string")
    try:
        size = len(value.encode("utf-8", errors="strict"))
    except UnicodeError:
        _fail("SCHEMA", "strings must contain Unicode scalars")
    if (nonempty and not size) or (maximum is not None and size > maximum):
        _fail("SCHEMA", "string length outside permitted bounds")


def _identifier(value: Any) -> None:
    if not isinstance(value, str) or IDENTIFIER.fullmatch(value) is None:
        _fail("SCHEMA", "invalid public identifier")


def _digest(value: Any) -> None:
    if not isinstance(value, str) or DIGEST.fullmatch(value) is None:
        _fail("SCHEMA", "invalid interface digest")


def _object(value: Any, keys: set[str] | None = None, code: str = "SCHEMA") -> dict:
    if not isinstance(value, dict) or any(not isinstance(key, str) for key in value):
        _fail(code, "expected object")
    if keys is not None and set(value) != keys:
        _fail(code, "object fields do not match contract")
    return value


def _array(value: Any, nonempty: bool = False) -> list:
    if not isinstance(value, list) or (nonempty and not value):
        _fail("SCHEMA", "expected array" + (" with at least one item" if nonempty else ""))
    return value


def _json_bounds(value: Any) -> None:
    stack = [(value, 0)]
    count = 0
    while stack:
        item, depth = stack.pop()
        count += 1
        if depth > MAX_JSON_DEPTH or count > MAX_NODES:
            _fail("LIMIT", "JSON depth or node budget exceeded")
        if isinstance(item, dict):
            for key, child in item.items():
                _string(key)
                stack.append((child, depth + 1))
        elif isinstance(item, list):
            stack.extend((child, depth + 1) for child in item)
        elif isinstance(item, str):
            _string(item)
        elif type(item) is int:
            if abs(item) > MAX_SAFE_INTEGER:
                _fail("SCHEMA", "JSON numbers must be safe integers")
        elif item is not None and type(item) is not bool:
            _fail("SCHEMA", "unsupported JSON value or nonintegral number")


def _pairs(items: list[tuple[str, Any]]) -> dict:
    result = {}
    for key, value in items:
        if key in result:
            _fail("FRAME", "duplicate JSON object key")
        result[key] = value
    return result


def _invalid_constant(_value: str) -> None:
    _fail("FRAME", "nonfinite JSON number")


def _exact_number(token: str) -> int:
    try:
        coefficient = re.split(r"[eE]", token, maxsplit=1)[0].replace("-", "").replace(".", "")
        if not coefficient.strip("0"):
            return 0
        number = Decimal(token)
        if (not number.is_finite() or not -MAX_SAFE_INTEGER <= number <= MAX_SAFE_INTEGER
                or number != number.to_integral_value()):
            _fail("SCHEMA", "JSON numbers must be exact safe integers")
        return int(number)
    except (DecimalException, OverflowError, ValueError):
        _fail("SCHEMA", "JSON number exceeds supported bounds")


def _parse_json(data: bytes) -> Any:
    try:
        value = json.loads(data.decode("utf-8", errors="strict"),
                           object_pairs_hook=_pairs, parse_constant=_invalid_constant,
                           parse_float=_exact_number, parse_int=_exact_number)
    except (UnicodeError, json.JSONDecodeError, RecursionError, ValueError) as exc:
        if isinstance(exc, ProtocolError):
            raise
        _fail("FRAME", "invalid UTF-8 JSON")
    _json_bounds(value)
    return value


def parse_frame(data: bytes) -> dict:
    """Parse exactly one LF-terminated frame, checking bytes before decoding."""
    if not isinstance(data, bytes):
        _fail("FRAME", "frame must be bytes")
    if len(data) > MAX_FRAME_BYTES + 1:
        _fail("LIMIT", "frame byte budget exceeded")
    if not data.endswith(b"\n") or b"\n" in data[:-1] or b"\r" in data:
        _fail("FRAME", "frame must have one LF terminator and no literal CR")
    if not data[:-1].strip():
        _fail("FRAME", "empty frame")
    return _object(_parse_json(data[:-1]))


def encode_frame(message: dict) -> bytes:
    """Serialize a bounded object. Call operation-specific validation first."""
    _object(message)
    _json_bounds(message)
    payload = json.dumps(message, ensure_ascii=False, separators=(",", ":"),
                         allow_nan=False).encode("utf-8")
    if len(payload) > MAX_FRAME_BYTES:
        _fail("LIMIT", "frame byte budget exceeded")
    return payload + b"\n"


def _request_id(value: Any) -> None:
    if type(value) is not int or not 1 <= value <= MAX_SAFE_INTEGER:
        _fail("SCHEMA", "request ID must be a positive safe integer")


def validate_request(message: Any) -> dict:
    _json_bounds(message)
    _object(message)
    if type(message.get("v")) is not int or message["v"] != 1:
        _fail("SCHEMA", "unsupported protocol version")
    _request_id(message.get("id"))
    op = message.get("op")
    fields = {"hello": {"interfaceDigest", "runtime"}, "create": set(),
              "invoke": {"action", "inputs"}, "observe": set(),
              "dispose": set(), "cancel": {"requestId"}}
    if not isinstance(op, str) or op not in fields:
        _fail("SCHEMA", "unknown operation")
    _object(message, {"v", "id", "op"} | fields[op])
    if op == "hello":
        _digest(message["interfaceDigest"])
        _identifier(message["runtime"])
    elif op == "invoke":
        _identifier(message["action"])
        _object(message["inputs"])
        for name in message["inputs"]:
            _identifier(name)
    elif op == "cancel":
        _request_id(message["requestId"])
        if message["requestId"] >= message["id"]:
            _fail("SCHEMA", "cancel target must precede its request ID")
    return message


def validate_response(message: Any) -> dict:
    _json_bounds(message)
    _object(message)
    if type(message.get("v")) is not int or message["v"] != 1:
        _fail("SCHEMA", "unsupported protocol version")
    _request_id(message.get("id"))
    if type(message.get("ok")) is not bool:
        _fail("SCHEMA", "response ok must be boolean")
    if message["ok"]:
        _object(message, {"v", "id", "ok", "result"})
    else:
        _object(message, {"v", "id", "ok", "error"})
        error = _object(message["error"], {"code", "message"})
        if not isinstance(error["code"], str) or ERROR_CODE.fullmatch(error["code"]) is None:
            _fail("SCHEMA", "invalid error code")
        _string(error["message"], 1024)
    return message


def validate_result(request: dict, response: dict, manifest: dict) -> Any:
    """Check correlation and success result shape; return result, or error dict."""
    validate_request(request)
    validate_response(response)
    if response["id"] != request["id"]:
        _fail("SCHEMA", "uncorrelated response")
    if not response["ok"]:
        return response["error"]
    result = response["result"]
    if request["op"] == "hello":
        _object(result, {"interfaceDigest", "runtime"})
        if (result["interfaceDigest"] != request["interfaceDigest"] or
                result["runtime"] != request["runtime"]):
            _fail("HANDSHAKE", "worker identity does not match requested interface/runtime")
    elif request["op"] == "observe":
        validate_observations(manifest, result)
    elif result is not None:
        _fail("SCHEMA", "operation result must be null")
    return result


def _unique(items: list, field: str, label: str) -> None:
    found = set()
    for item in items:
        value = item[field]
        if value in found:
            _fail("SCHEMA", "duplicate " + label)
        found.add(value)


def _type(shape: Any, depth: int = 0) -> None:
    if depth > MAX_TYPE_DEPTH:
        _fail("LIMIT", "type depth budget exceeded")
    _object(shape)
    kind = shape.get("kind")
    if kind in ("int", "bool", "str", "null"):
        _object(shape, {"kind"})
    elif kind in ("set", "seq"):
        _object(shape, {"kind", "element"})
        _type(shape["element"], depth + 1)
    elif kind == "tuple":
        _object(shape, {"kind", "elements"})
        for item in _array(shape["elements"]):
            _type(item, depth + 1)
    elif kind == "record":
        _object(shape, {"kind", "fields"})
        fields = _array(shape["fields"])
        for item in fields:
            _object(item, {"wireName", "type"})
            _string(item["wireName"], 128, True)
            _type(item["type"], depth + 1)
        _unique(fields, "wireName", "record field")
    elif kind == "map":
        _object(shape, {"kind", "key", "value"})
        if shape["key"] != {"kind": "str"}:
            _fail("SCHEMA", "portable maps require string keys")
        _type(shape["key"], depth + 1)
        _type(shape["value"], depth + 1)
    elif kind == "variant":
        _object(shape, {"kind", "cases"})
        cases = _array(shape["cases"], True)
        for item in cases:
            _object(item, {"tag", "payload"})
            _string(item["tag"], 128, True)
            _type(item["payload"], depth + 1)
        _unique(cases, "tag", "variant tag")
    else:
        _fail("SCHEMA", "unsupported portable type")


def validate_type(shape: Any) -> dict:
    _json_bounds(shape)
    _type(shape)
    return shape


def validate_manifest(manifest: Any) -> dict:
    _json_bounds(manifest)
    _object(manifest, {"schema", "interfaceDigest", "initializers", "actions", "observations"})
    if manifest["schema"] != "mirrorgate.port/v1":
        _fail("SCHEMA", "unsupported public manifest schema")
    _digest(manifest["interfaceDigest"])
    initializers = _array(manifest["initializers"], True)
    actions = _array(manifest["actions"])
    observations = _array(manifest["observations"], True)
    for action in initializers + actions:
        _object(action, {"id", "inputs"})
        _identifier(action["id"])
        inputs = _array(action["inputs"])
        for item in inputs:
            _object(item, {"id", "type"})
            _identifier(item["id"])
            _type(item["type"])
        _unique(inputs, "id", "input ID")
    _unique(initializers + actions, "id", "operation ID")
    for item in observations:
        _object(item, {"id", "type"})
        _identifier(item["id"])
        _type(item["type"])
    _unique(observations, "id", "observation ID")
    if len(json.dumps(manifest, ensure_ascii=False, separators=(",", ":")).encode("utf-8")) > MAX_MANIFEST_BYTES:
        _fail("LIMIT", "manifest byte budget exceeded")
    return manifest


def load_manifest(path: str | Path) -> dict:
    with Path(path).open("rb") as stream:
        data = stream.read(MAX_MANIFEST_BYTES + 1)
    if len(data) > MAX_MANIFEST_BYTES:
        _fail("LIMIT", "manifest byte budget exceeded")
    return validate_manifest(_parse_json(data))


def _value(shape: dict, value: Any, depth: int = 0) -> Any:
    """Validate and return a hashable semantic key for set/map duplicate checks."""
    if depth > MAX_VALUE_DEPTH:
        _fail("LIMIT", "value depth budget exceeded")
    kind = shape["kind"]
    if kind == "int":
        _object(value, {"#bigint"}, "VALUE")
        number = value["#bigint"]
        if not isinstance(number, str) or INTEGER.fullmatch(number) is None:
            _fail("VALUE", "integer must be a canonical decimal string")
        return (kind, number)
    if kind == "bool":
        if type(value) is not bool:
            _fail("VALUE", "expected boolean")
        return (kind, value)
    if kind == "str":
        if not isinstance(value, str):
            _fail("VALUE", "expected string")
        return (kind, value)
    if kind == "null":
        if value is not None:
            _fail("VALUE", "expected null")
        return (kind,)
    if kind in ("set", "seq", "tuple"):
        values = value
        if kind != "seq":
            marker = "#set" if kind == "set" else "#tup"
            _object(value, {marker}, "VALUE")
            values = value[marker]
        if not isinstance(values, list):
            _fail("VALUE", "expected collection array")
        if kind == "tuple":
            if len(values) != len(shape["elements"]):
                _fail("VALUE", "tuple arity does not match")
            keys = tuple(_value(t, v, depth + 1) for t, v in zip(shape["elements"], values))
        else:
            keys = tuple(_value(shape["element"], v, depth + 1) for v in values)
        if kind == "set":
            unique = frozenset(keys)
            if len(unique) != len(keys):
                _fail("VALUE", "duplicate set member")
            return (kind, unique)
        return (kind, keys)
    if kind == "record":
        fields = shape["fields"]
        _object(value, {item["wireName"] for item in fields}, "VALUE")
        return (kind, frozenset((item["wireName"], _value(item["type"], value[item["wireName"]], depth + 1))
                                for item in fields))
    if kind == "map":
        _object(value, {"#map"}, "VALUE")
        entries = value["#map"]
        if not isinstance(entries, list):
            _fail("VALUE", "expected map entry array")
        seen = set()
        keys = []
        for entry in entries:
            if not isinstance(entry, list) or len(entry) != 2 or not isinstance(entry[0], str):
                _fail("VALUE", "expected string-key map pair")
            if entry[0] in seen:
                _fail("VALUE", "duplicate map key")
            seen.add(entry[0])
            keys.append((entry[0], _value(shape["value"], entry[1], depth + 1)))
        return (kind, frozenset(keys))
    if kind == "variant":
        _object(value, {"tag", "value"}, "VALUE")
        for case in shape["cases"]:
            if value["tag"] == case["tag"]:
                return (kind, case["tag"], _value(case["payload"], value["value"], depth + 1))
        _fail("VALUE", "unknown variant tag")
    _fail("SCHEMA", "unsupported portable type")


def validate_value(shape: Any, value: Any) -> Any:
    validate_type(shape)
    _json_bounds(value)
    _value(shape, value)
    return value


def operation(manifest: dict, action: str) -> dict:
    for item in manifest["initializers"] + manifest["actions"]:
        if item["id"] == action:
            return item
    _fail("VALUE", "undeclared operation")


def validate_inputs(manifest: dict, action: str, inputs: Any) -> dict:
    selected = operation(manifest, action)
    _json_bounds(inputs)
    _object(inputs, {item["id"] for item in selected["inputs"]}, "VALUE")
    for item in selected["inputs"]:
        _value(item["type"], inputs[item["id"]])
    return inputs


def validate_observations(manifest: dict, values: Any) -> dict:
    _json_bounds(values)
    _object(values, {item["id"] for item in manifest["observations"]}, "VALUE")
    for item in manifest["observations"]:
        _value(item["type"], values[item["id"]])
    return values
